test/metabase/query_processor_test/string_extracts_test.clj

```diff
@@ -70,9 +71,30 @@
 
 (deftest ^:parallel test-concat
   (mt/test-drivers (mt/normal-drivers-with-feature :expressions)
-    (is (= "foobar" (test-string-extract [:concat "foo" "bar"])))
+    (testing "Does concat work with 2 strings"
+      (is (= "foobar" (test-string-extract [:concat "foo" "bar"]))))
     (testing "Does concat work with >2 args"
-      (is (= "foobar" (test-string-extract [:concat "f" "o" "o" "b" "a" "r"]))))))
+      (is (= "foobar" (test-string-extract [:concat "f" "o" "o" "b" "a" "r"]))))
+    (testing "Does concat work with nested concat expressions"
+      (is (= "foobar" (test-string-extract [:concat [:concat "f" "o" "o"] [:concat "b" "a" "r"]]))))))
+
+(defmethod driver/database-supports? [::driver/driver ::concat-non-string-args]
+  [_driver _feature _database]
+  true)
+
+;; These drivers do not support concat with non-string args
```

Comment:
Is this the right way to exclude these drivers from this test? Ad-hoc, file-local features like this seem to be the most common idiom in our tests, but occasionally we add a key like `:test/dynamic-dataset-loading` or `:test/jvm-timezone-setting` to `metabase.driver.features`. And then some features are a middle ground between the two, like `:test/timestamptz-type` or `:test/time-type` which are `:test`-prefixed and implemented in a common place (`metabase.test.data.interface`), but not added to `metabase.driver.features`.

Pinging @lbrdnk as I seem to recall seeing a comment from him related to this topic in a gh review somewhere, but can't find it now...

Reply:
I believe you mean [this](https://github.com/metabase/metabase/pull/41384#issuecomment-2289243742) reply from Cam to my review or the original comment.

Based on that, the general features go into the `metabase.driver/features` and ad-hoc, ie. disabling just one test, should live right next to it.

With regards to namespacing, I used `:test/...` in the `dynamic-dataset-loading` to make it clear that this is related _only_ to our tests.

I think we are still in the process of finding the universal approach. Current model is fine, but as we define more and more test features it may become hard to find the real features among those. Either the product features or dbms quirks vs features defined just for the purposes of disabling some test/s.

That said, in this situation, I'd define the feat in the general namespace. But also I'm not saying it is the only right answer :)

Reply:
That's the comment I was remembering, thanks! Really good comment btw, all good callouts.

> That said, in this situation, I'd define the feat in the general namespace

To clarify, you mean naming the feature `:test/concat-non-string-args` but not adding it to `driver/features` right? Similar to, e.g. `:test/time-type`?

Personally, I think I lean towards leaving it ad-hoc since it's used in a single test and namespace, but could be convinced otherwise. Wdyt?

Reply:
> Personally, I think I lean towards leaving it ad-hoc since it's used in a single test and namespace...

You are right. I've edited my comment realizing that's the case. I think it is fine the way it is.